mmeval/metrics/perplexity.py

```diff
@@ -0,0 +1,294 @@
+
+        Args:
+            prediction (paddle.Tensor | Sequence): Prediction from the model.
+                Same as ``self.add``.
+            target (paddle.Tensor | Sequence): The ground truth labels. Same as
+                ``self.add``.
+
+        Returns:
+            Tuple (float, int): include the value of the total and count.
+        """
+        probs = paddle.nn.functional.softmax(prediction, axis=1)
+        if self.ignore_labels is not None:
+            mask = paddle.zeros_like(target, dtype=bool)
+            for ignore_label in self.ignore_labels:
+                mask_temp = target.equal(ignore_label)
+                mask = (mask == mask_temp)
+            replace = paddle.zeros_like(target)
+            target = paddle.where(mask, replace, target)
+            probs = probs.index_select(axis=1, index=target).diagonal()[~mask]
+        else:
+            probs = probs.index_select(axis=1, index=target).diagonal()
+        total = -probs.log().sum(axis=0)
+        count = paddle.to_tensor(probs.shape[0])
+        result = (total.item(), count.item())
+        return result
+
+    @dispatch
+    def _compute_perplexity(self, prediction: np.ndarray,
+                            target: np.ndarray) -> Tuple[float, int]:
+        """Compute the perplexity with NumPy.
+
+        Args:
+            prediction (np.ndarray | Sequence): Prediction from the model.
+                Same as ``self.add``.
+            target (np.ndarray | Sequence): The ground truth labels. Same as
+                ``self.add``.
+
+        Returns:
+            Tuple (float, int): include the value of the total and count.
+        """
+        probs = softmax(prediction)
+        if self.ignore_labels is not None:
+            mask = np.zeros_like(target, dtype=bool)
+            for ignore_label in self.ignore_labels:
+                mask_temp = np.equal(target, ignore_label)
+                mask = (mask == mask_temp)
+            target = np.ma.array(target, mask=mask, fill_value=0).filled()
+            probs = np.take(probs, target, axis=1).diagonal()[~mask]
+        else:
+            probs = np.take(probs, target, axis=1).diagonal()
+        total = -np.sum(np.log(probs))
+        count = probs.size
+        result = (total.item(), count)
+        return result
+
+    def compute_metric(self, results: List[Tuple[float, int]]) \
+            -> Dict[str, float]:
+        """Compute the perplexity metric.
+
+        This method would be invoked in ``BaseMetric.compute`` after
+        distributed synchronization.
+
+        Args:
+            results (list): A list that consisting the total and count. This
+                list has already been synced across all ranks.
+
+        Returns:
+            Dict[str, float]: The computed perplexity metric.
+        """
+        total = 0.0
+        count = 0
+        for result in results:
+            total += result[0]
+            count += result[1]
+        output = np.exp(total / count)
```

Comment:
Should we handle the case where count = 0?

Reply:
Yes, there should be another insurance to prevent mistakes. I'll modify it